ssl/statem/extensions_clnt.c

```diff
@@ -1584,6 +1586,28 @@ int tls_parse_stoc_alpn(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
         SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_R_BAD_EXTENSION);
         return 0;
     }
+
+    /* It must be a protocol that we sent */
+    if (!PACKET_buf_init(&confpkt, s->ext.alpn, s->ext.alpn_len)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    while (PACKET_get_length_prefixed_1(&confpkt, &protpkt)) {
+        if (PACKET_remaining(&protpkt) != len)
+            continue;
```

Comment:
I may be way off base here, but I _think_ this could cause a DoS.  If I'm reading this right, we're doing the following:

1) Getting a vector from the inbound packet
2) Ensuring that there is at least len bytes available in the vector read from the packet.

If a malicious sender creates a packet for which the protocol list is nothing but a series of 0 values, won't this loop treat every byte in that list as a length vector, causing us to iterate many hundreds of times through the continue statment on line 1597.  That could be many 1000's of iterations per handshake (unsure how big this extension can be).

Maybe not a DoS, but is it worth checking for malformed length values here?

Reply:
Ok, I think what you're saying is this packet is generated and parsed locally, so theres no opportunity for a remote user to send an invalid packet.  if so, then yes, I agree, its ok

Reply:
Essentially, yes. Although the term "packet" is a bit misleading. We're using the PACKET API here - but what we're parsing is an octet string set via a config file, or via a previous API call. It's not a network packet.